src/common/snippets/docs/snippets_design_guide.md

```diff
@@ -1,301 +1,465 @@
 
-Parameters for a generated snippet are split into schedule invariant and schedule dependent. Schedule-invariant parameters include pointers to input/output tensors and strides for each of them with the same rank as the scheduling domain.
+### Generator
+The main purpose of the `Generator` is to emit executable code. The code emission process could be divided into two stages: `Preparation` that could be shared between several targets, and target-specific `Emission`. The `ov::snippets::Generator` class provides a target-independent interface to the generation process and performs the `Preparation` stage. It also stores the `TargetMachine` instance, which is responsible for the `Emission` stage. Let's discuss the target-independent stage first.
 
-### Diagnostics
+The `Preparation` consists of low-level target-independent transformations needed to prepare the `IR` for code generation. There are currently two such transformations:
+1. `AssignRegisters` assigns abstract registers to `Expressions` based on their data dependencies. The easiest way to think about the assignment logic is that a register is assigned to every `PortConnector` to ensure appropriate data propagation. However, every `Expression` needs to know the assigned registers, so they are stored is the `PortDescriptors` (and could be obtained via `get_reg()`). Consequently, all the ports connected to the same `PortConnector` will have the same register in their `PortDescriptors`. `AssignRegisters` also supports register re-utilization, so if all `ExpressionPorts` connected to this `PortConnector` are evaluated, then the corresponding register may be reused by other `PortConnector`. In other words, when all the `Expressions` that required input data in a certain register are evaluated, the register may be reused to hold another `Expression's` output. `AssignRegisters` also supports two types of registers: general-purpose and vector ones. Different types of registers are managed and assigned independently, and a particular register type required by an `Expression` is provided by the `ov::snippets::Generator` (or a derived generator for target-specific `Ops`).  
+2. `InsertTailLoop` injects tail-processing section after a loop body if needed. Note that every loop has two parameters that specify how its body is evaluated: `work_amount` and `increment` The `work_amount` indicates how much of the data needs to be processed, it often equals to the dimension's size the loop is working on. The `increment` defines how many data entries are processed on every loop iteration (it usually equals to vector size for the innermost loops of elementwise subgraph). So if a loop's `work_amount` is not evenly divisible by its `increment`, it means that a tail processing is required. `InsertTailLoop` duplicates the body of such a loop, rescales pointer increments and load/store masks appropriately, and injects these `Ops` immediately after the processed loop.
```

Comment:
Maybe we can describe `Loop` work before as the main part of opset? I mean to explain that to emulate loops there are two ops `LoopBegin` and `LoopEnd`. The loop has the parameters for cycle execution: `work_amount` (loop trip count), `increment` (step), etc. And add the diagram from `Linear IR`

What do you think?
I hope it would be really helpful for new developers to understand how to work with loops

Reply:
Totally agree. Moreover, all Snippets-specific operations should be covered, especially Kernel, LoopBegin, LoopEnd, MemoryAcess class, etc. 
I've created a ticket for this: 111895